index.html

```diff
@@ -1453,33 +1471,143 @@ <h3>The <dfn data-dfn-for="">Observable</dfn> interface</h3>
+          <span class="rfc2119-assertion" id="scripting-sec-interface-min">
+              Therefore the WoT Runtime SHOULD only expose a minimal set of interfaces to a WoT script based
+               on its intended functionality.
+          </span>            
+      </li>
+      <li>If the WoT runtime supports post-manufacturing provisioning or update of WoT scripts, WoT runtime
+          or any related data (including security credentials), it can be a major attack vector. An attacker
+          can try to modify any above described part during the update or provisioning process or simply
+          provision attacker's code and data directly. 
+          <span class="rfc2119-assertion" id="scripting-sec-update">
+              Therefore, if WoT Runtime supports post-manufacturing provisioning or update of WoT scripts, WoT
+              runtime or any related data, such operations SHOULD be done in a secure fashion.
+          </span>
+          A set of recommendations for secure update and post-manufacturing provisioning can be found in
+          [[!WOT-SECURITY-BEST-PRACTICES]].
+      </li>
+      <li>Typically the WoT runtime needs to store the security credentials that are provisioned to a WoT
+      device to operate in WoT network. The confidentiality or integrity of these credentials should not 
+      be compromised. 
+          <span class="rfc2119-assertion" id="scripting-sec-creds1">
+              Therefore the WoT runtime SHOULD securely store the provisioned security credentials,
+              guaranteeing their integrity and confidentiality. 
+          </span>
+          <span class="rfc2119-assertion" id="scripting-sec-creds2">
+            In case there are more than one tenant on a single WoT-enabled device, a WoT Runtime SHOULD 
+            guarantee isolation of each tenant provisioned security credentials. 
+          </span>
+          Additionally, in order to minimize a risk that provisioned security credentials get compromised,
+          the WoT runtime should not have any way for WoT scripts to query these credentials.
+          <span class="rfc2119-assertion" id="scripting-sec-creds3">
+            Therefore, the WoT Runtime SHOULD NOT expose any API for WoT scripts to query the provisioned
+            security credentials. 
+          </span>
+      </li>
+   </ul>
+  </p>
+
+ <p>Some additional specific recommendations relevant for WoT script developers:
+  <ul>
+        <li>A typical way to compromise any process is to send it a corrupted input via one of the exposed
+           interfaces. 
+          <span class="rfc2119-assertion" id="scripting-sec-inputs">
+           Therefore developers SHOULD perform validation on all WoT script inputs, including 
+           <a href="https://en.wikipedia.org/wiki/Fuzzing">fuzzing</a>.
+          </span>
+            There are many tool and techniques in existence to do such validation. 
+            More details can be found in [[!WOT-SECURITY-TESTING]].
+
```

Comment:
I am not sure we should put the burden solely on script developers. We should identify the inputs that are possible via scripts and insert normative and informative instructions to WoT Runtime implementors on what and how to check.

In our case I think the TD (provided to `produce()`) is one of the possible threats. Fortunately the TD document should contain normative instructions on TD parsing, and we could hopefully link those from the Scripting spec.

Other input comes from protocol replies, and that is clearly the WoT Runtimes responsibility.

Reply:
Sure, more we can provide is better. But there is always a part of negative testing that cannot be specified, such as fuzzing. You can build a custom fuzzer to take into account the TD format, but its inputs are not predefined.